middleware/kubernetes/kubernetes.go

```diff
@@ -167,31 +170,46 @@ func (k *Kubernetes) getZoneForName(name string) (string, []string) {
 			break
 		}
 	}
+	if strings.HasPrefix(serviceSegments[0], "_") {
+		if strings.HasPrefix(serviceSegments[1], "_") {
+			srvPort = serviceSegments[0]
+			srvProtocol = serviceSegments[1]
+			serviceSegments = serviceSegments[2:len(serviceSegments)]
+		} else {
+			srvProtocol = serviceSegments[0]
+			serviceSegments = serviceSegments[1:len(serviceSegments)]
+		}
+	}
+
+	return srvPort, srvProtocol, serviceSegments, zone
+}
 
-	return zone, serviceSegments
+func parseServiceSegments(serviceSegments []string) (string, string, string, string) {
+	serviceSegments, typeName := pullOffFromRight(serviceSegments)
+	serviceSegments, namespace := pullOffFromRight(serviceSegments)
+	serviceSegments, serviceName := pullOffFromRight(serviceSegments)
+	serviceSegments, endpointName := pullOffFromRight(serviceSegments)
+
+	return endpointName, serviceName, namespace, typeName
 }
 
 // Records looks up services in kubernetes. If exact is true, it will lookup
 // just this name. This is used when find matches when completing SRV lookups
 // for instance.
 func (k *Kubernetes) Records(name string, exact bool) ([]msg.Service, error) {
-	var (
-		serviceName string
-		namespace   string
-		typeName    string
-	)
-
-	zone, serviceSegments := k.getZoneForName(name)
-	if len(serviceSegments) < 3 {
+	var records []msg.Service
+
+	srvPort, srvProtocol, serviceSegments, zone := k.getPrefixesAndZoneForName(name)
+	endpointName, serviceName, namespace, typeName := parseServiceSegments(serviceSegments)
+
+	// If this is a SRV request, exit if query string doesnt contain the port/protocol
+	if k.RequestType == "SRV" && (srvPort == "" || srvProtocol == "") {
```

Comment:
In the etcd case we don't care about the request type (qtype) and this level. I think it is wise not to pull the logic into this function and just let it dig up the records and then let the caller just check if the right stuff is there. You'll then also don't need the hack of puting requestType in the receiver, instead of extending the methods parameter list.